alephium/contracts/governance.ral

```diff
@@ -0,0 +1,202 @@
+    fn getGuardiansInfo(guardianSetIndex: U256) -> ByteVec {
+        if (guardianSetIndex == guardianSetIndexes[1]) {
+            return guardianSets[1]
+        }
+        if (guardianSetIndex == guardianSetIndexes[0]) {
+            assert!(blockTimeStamp!() <= previousGuardianSetExpirationTime, ErrorCodes.GuardianSetExpired)
+            return guardianSets[0]
+        }
+        panic!(ErrorCodes.InvalidGuardianSetIndex)
+    }
+
+    pub fn submitContractUpgrade(vaa: ByteVec) -> () {
+        let (targetChainId, payload) = parseAndVerifyGovernanceVAA(vaa, ActionId.ContractUpgrade)
+        assert!(targetChainId == chainId, ErrorCodes.InvalidTargetChainId)
+        let contractCodeLength = u256From2Byte!(byteVecSlice!(payload, 33, 35))
+        let mut offset = 35 + contractCodeLength
+        let newContractCode = byteVecSlice!(payload, 35, offset)
+        let payloadLength = size!(payload)
+        if (payloadLength == offset) {
+            migrate!(newContractCode)
+        } else {
+            let prevStateHash = byteVecSlice!(payload, offset, offset + 32)
+            offset = offset + 32
+            let currentEncodedState = u256To32Byte!(receivedSequence)
+                ++ u256To32Byte!(messageFee)
+                ++ u256To32Byte!(guardianSetIndexes[0])
+                ++ u256To32Byte!(guardianSetIndexes[1])
+            assert!(prevStateHash == blake2b!(currentEncodedState), ErrorCodes.ContractStateMismatch)
+            let stateLength = u256From2Byte!(byteVecSlice!(payload, offset, offset + 2))
+            offset = offset + 2
+            let newContractState = byteVecSlice!(payload, offset, offset + stateLength)
+            offset = offset + stateLength
+            assert!(payloadLength == offset, ErrorCodes.InvalidMessageSize)
+            migrateWithFields!(newContractCode, newContractState)
+        }
+    }
+
+    fn updatePreviousGuardianSet() -> () {
+        guardianSets[0] = guardianSets[1]
+        guardianSetIndexes[0] = guardianSetIndexes[1]
+        previousGuardianSetExpirationTime = blockTimeStamp!() + GuardianSetExpireDuration
+    }
+
+    pub fn submitNewGuardianSet(vaa: ByteVec) -> () {
+        updatePreviousGuardianSet()
+        let (targetChainId, payload) = parseAndVerifyGovernanceVAA(vaa, ActionId.NewGuardianSet)
+        assert!(targetChainId == chainId || targetChainId == 0, ErrorCodes.InvalidTargetChainId)
+        let newGuardianSetIndex = u256From4Byte!(byteVecSlice!(payload, 33, 37))
+        assert!(newGuardianSetIndex == guardianSetIndexes[1] + 1, ErrorCodes.InvalidGuardianSetIndex)
+        let newGuardianSetSize = u256From1Byte!(byteVecSlice!(payload, 37, 38))
+        assert!(newGuardianSetSize > 0, ErrorCodes.InvalidGuardianSetSize)
+        let payloadSize = 38 + newGuardianSetSize * 20
+        assert!(size!(payload) == payloadSize, ErrorCodes.InvalidMessageSize)
+        guardianSetIndexes[1] = newGuardianSetIndex
+        guardianSets[1] = byteVecSlice!(payload, 37, payloadSize)
```

Comment:
Should this be `guardianSets[1] = byteVecSlice!(payload, 38, payloadSize)` ?

Reply:
The encoding of `guardianSet` is: 1 + 20 * (the number of guardians), the first byte is the number of guardians, so we use 37 here.

Reply:
Awesome, thanks for the explanation :+1: